sdk/armcore/poller.go

```diff
@@ -6,958 +6,307 @@
 	if err != nil {
 		return nil, err
 	}
-	if p.pt.finalGetURL() == "" {
-		// we can end up in this situation if the async operation returns a 200
-		// with no polling URLs.  in that case return the response which should
-		// contain the JSON payload (only do this for successful terminal cases).
-		if lr := p.pt.latestResponse(); lr != nil && p.pt.hasSucceeded() {
-			result, err := p.handleResponse(lr, respType)
-			if err != nil {
-				return nil, err
-			}
-			return result, nil
-		}
-		return nil, errors.New("missing URL for retrieving result")
-	}
-	req, err := azcore.NewRequest(ctx, http.MethodGet, p.pt.finalGetURL())
+	resp, err := l.Pipeline.Do(req)
 	if err != nil {
+		// don't update the poller for failed requests
 		return nil, err
 	}
-	resp, err := pipeline.Do(req)
-	if err != nil {
+	defer resp.Body.Close()
+	if !lroStatusCodeValid(resp) {
+		// the LRO failed.  unmarshall the error and update state
+		l.err = l.eu(resp)
+		l.resp = nil
+		return nil, l.err
+	}
+	if err = l.lro.Update(resp); err != nil {
 		return nil, err
 	}
-	return p.handleResponse(resp, respType)
+	l.resp = resp
+	azcore.Log().Writef(azcore.LogLongRunningOperation, "Status %s", l.lro.Status())
+	if pollers.Failed(l.lro.Status()) {
+		l.err = l.eu(resp)
+		l.resp = nil
+		return nil, l.err
+	}
+	return l.resp.Response, nil
 }
 
-func (p *poller) ResumeToken() (string, error) {
-	if p.pt.hasTerminated() {
+// ResumeToken returns a token string that can be used to resume a poller that has not yet reached a terminal state.
+func (l *LROPoller) ResumeToken() (string, error) {
+	if l.Done() {
 		return "", errors.New("cannot create a ResumeToken from a poller in a terminal state")
 	}
-	js, err := json.Marshal(p.pt)
+	b, err := json.Marshal(l.lro)
 	if err != nil {
 		return "", err
 	}
-	return string(js), nil
+	return string(b), nil
 }
 
-func (p *poller) handleResponse(resp *azcore.Response, respType interface{}) (*http.Response, error) {
-	if resp.HasStatusCode(http.StatusNoContent) {
-		return resp.Response, nil
+// FinalResponse will perform a final GET request and return the final HTTP response for the polling
+// operation and unmarshall the content of the payload into the respType interface that is provided.
+func (l *LROPoller) FinalResponse(ctx context.Context, respType interface{}) (*http.Response, error) {
+	if !l.Done() {
+		return nil, errors.New("cannot return a final response from a poller in a non-terminal state")
 	}
-	if !resp.HasStatusCode(pollingCodes[:]...) {
-		return nil, p.pt.handleError(resp)
+	// if there's nothing to unmarshall into or no response body just return the final response
+	if respType == nil {
+		return l.resp.Response, nil
+	} else if l.resp.StatusCode == http.StatusNoContent || l.resp.ContentLength == 0 {
+		azcore.Log().Write(azcore.LogLongRunningOperation, "final response specifies a response type but no payload was received")
+		return l.resp.Response, nil
```

Comment:
This isnt an error right?

Reply:
Correct.  I wanted to log this in case customers ask why a poller returns an empty `Widget`.  This would indicate a swagger or service issue (i.e. the swagger says the operation returns a `Widget` but the endpoint doesn't).

Reply:
Ok yeah, it seemed strange to me that it wouldnt be an error if that's what's expected, but I get it being needed for some services that might have this issue.